website/source/docs/agent/telemetry.html.md

```diff
@@ -933,6 +933,12 @@ These metrics give insight into the health of the cluster as a whole.
     <td>queries</td>
     <td>counter</td>
   </tr>
+  <tr>
+    <td>`consul.catalog.service.query-tags.<service>.<tags>`</td>
+    <td>This increments for each catalog query for the given service with the given tags.</td>
```

Comment:
So if there are multiple tags like `tags=foo&tags=bar` will this result in separate increments for each, or in increment on a single (possibly new counter) `query-tags.<service>.foo,bar` or something? If it's the latter maybe we should document the format of the key or given an example?

Did we add this metric? (Yes, just checked the original PR). I thought we just changed some existing labelling. I guess I didn't pay enough attention to the original PR! Adding metrics with an unbounded cardinality is actually problematic, we should consider reverting that part...

What was the rationale for wanting those? The argument against is that most metrics systems have a finite capacity of distinct series they can collect and having labels like this where any misbehaving client (e.g. a confused developer) can try a few permutations of tags on a few agents and cause thousands of new series makes capacity planning hard and can cause issues. Adding new metrics that are super linear in cluster size always feels a bit problematic to me unless they require opt-in in config so it can be a conscious decision from operators.

At worst they can cause an easy DOS vector for metrics systems - malicious code anywhere on the network just has to issue queries with different tag combinations and can generate as many new series as they want which will easily overload metrics systems/balloon memory etc.

For balance, there are almost certainly places in Consul that are already susceptible to this so it's potentially not adding anything worse than what we have, but I do like to be thoughtful about that going forward. See #4042 for an example of another great PR that I think is hampered by causing too much churn in metrics.

Happy to discuss though - if the value is really high we can consider leaving this as it is, I just worry about the risk compared to the potential benefit?

Reply:
A query with multiple tags will emit a different metric with all of the tag values specified. To be clear, there is still only a single metric emitted for the request, so there isn't a dramatic increase in the number of metrics emitted. 

To address the proliferation potential of time series, I now sort the tags before creating the metric. This means that these two queries will be emitted as the same metric: `?tag=foo&tag=bar`, `?tag=bar&tag=foo`.

The format for the output will be determined by the specific metric provider, based on the implementation from the [metrics library](https://github.com/armon/go-metrics/tree/master).